File: src/main/java/com/mrpowergamerbr/loritta/frontend/views/DonateView.java
package com.mrpowergamerbr.loritta.frontend.views;

import java.io.PrintWriter;
import java.io.StringWriter;
import com.mitchellbosecke.pebble.error.PebbleException;
import com.mitchellbosecke.pebble.template.PebbleTemplate;
import com.mrpowergamerbr.loritta.Loritta;
import com.mrpowergamerbr.loritta.frontend.LorittaWebsite;
import com.mrpowergamerbr.loritta.frontend.utils.RenderContext;
import com.mrpowergamerbr.temmiemercadopago.mp.Payment;
import com.mrpowergamerbr.temmiemercadopago.mp.TemmieItem;
import com.mrpowergamerbr.temmiemercadopago.mp.request.PaymentRequest;

public class DonateView {

	public static Object render(RenderContext context) {
		try {
			if (context.request().param("whodonated").isSet() && context.request().param("grana").isSet()) {
				try {
					String whoDonated = context.request().param("whodonated").value();
					double grana = Double.parseDouble(context.request().param("grana").value().replace(",", ".").replace("R$", "").replace("$", ""));

					grana = Math.max(0.01, grana);
					grana = Math.min(1000, grana);
					Payment payment = Loritta.getTemmieMercadoPago().generatePayment(PaymentRequest.builder()
							.addItem(TemmieItem.builder()
									.title("Doação para a Loritta - " + whoDonated)
									.quantity(1)
									.currencyId("BRL")
									.unitPrice(grana)
									.build()).build());
					
					context.response().redirect(payment.getInitPoint());
				} catch (Throwable e) {}
			}
			PebbleTemplate template = LorittaWebsite.getEngine().getTemplate("donate.html");
			return template;
		} catch (PebbleException e) {
			StringWriter sw = new StringWriter();
			PrintWriter pw = new PrintWriter(sw);
			e.printStackTrace(pw);
			return e.toString();
		}
	}
}

File: src/main/java/com/mrpowergamerbr/loritta/frontend/views/subviews/DonateView.kt
package com.mrpowergamerbr.loritta.frontend.views.subviews

import com.mrpowergamerbr.loritta.Loritta
import com.mrpowergamerbr.loritta.frontend.evaluate
import com.mrpowergamerbr.loritta.utils.lorittaShards
import com.mrpowergamerbr.temmiemercadopago.mp.TemmieItem
import com.mrpowergamerbr.temmiemercadopago.mp.request.PaymentRequest
import org.jooby.Request
import org.jooby.Response

class DonateView : AbstractView() {
	override fun handleRender(req: Request, res: Response, variables: MutableMap<String, Any?>): Boolean {
		return req.path() == "/donate"
	}

	override fun render(req: Request, res: Response, variables: MutableMap<String, Any?>): String {
		if (req.param("who-donated").isSet && req.param("grana").isSet) {
			val whoDonated = req.param("who-donated").value()
			var grana = req.param("grana").value().replace(",", ".").replace("R$", "").replace("$", "").toDoubleOrNull()

			if (grana != null) {
				grana = Math.max(0.01, grana)
				grana = Math.min(1000.0, grana)
				val payment = Loritta.temmieMercadoPago!!.generatePayment(PaymentRequest.builder()
						.addItem(TemmieItem.builder()
								.title("Doação para a Loritta - " + whoDonated)
								.quantity(1)
								.currencyId("BRL")
								.unitPrice(grana)
								.build()).build())

				res.redirect(payment.initPoint)
				return "Redirecionando..."
			}
		}
		val lorittaGuild = lorittaShards.getGuildById("297732013006389252")!!
		val rolePatreons = lorittaGuild.getRoleById("364201981016801281") // Pagadores de Aluguel
		val roleDonators = lorittaGuild.getRoleById("334711262262853642") // Doadores

		val patreons = lorittaGuild.getMembersWithRoles(rolePatreons)
		val donators = lorittaGuild.getMembersWithRoles(roleDonators)

		variables["patreons"] = patreons
		variables["donators"] = donators
		return evaluate("donate.html", variables)
	}
}